apps/prairielearn/src/pages/instructorStudents/components/DownloadButton.tsx

```diff
@@ -0,0 +1,85 @@
+import type { Table } from '@tanstack/react-table';
+
+import { downloadAsCSV, downloadAsJSON } from '../../../lib/client/downloads.js';
+import type { StudentRow } from '../instructorStudents.shared.js';
+
+export function DownloadButton({
+  students,
+  table,
+}: {
+  students: StudentRow[];
+  table: Table<StudentRow>;
+}) {
+  function downloadStudentsCSV(students: StudentRow[], filename: string): void {
+    const rows = students.map((student) => [
+      student.uid,
+      student.name,
+      student.email,
+      student.created_at
+        ? new Date(student.created_at).toISOString().replace('T', ' ').split('.')[0]
+        : '',
+    ]);
+    downloadAsCSV(['UID', 'Name', 'Email', 'Enrolled At'], rows, filename);
+  }
+  return (
+    <div class="btn-group">
+      <button
+        type="button"
+        data-bs-toggle="dropdown"
+        aria-expanded="false"
+        class="btn btn-outline-light btn-sm dropdown-toggle"
+      >
+        <i aria-hidden="true" class="px-2 fa fa-download"></i>
+        Download
+      </button>
+      <ul class="dropdown-menu">
+        <li>
+          <button
+            class="dropdown-item"
+            type="button"
+            onClick={() => downloadAsJSON(students, 'students.csv')}
```

Comment:
Elsewhere, we include the course and course instance information in filenames when downloading. For instance, downloading the gradebook CSV produces a file called `XC_101_SectionA_gradebook.csv`. This is important for instructors who might be juggling multiple courses or course instances.

Reply:
Despite the fact that they're nullable in the database, all non-deleted courses and course instances must by definition have non-null `short_name` values. Elsewhere where we deal with this, we use runtime assertions.

`node:assert` won't work here since this wants to be bundled on the client. We can either roll our own assert, or compute the prefix on the server where we can use `node:assert`, which would also have the nice property of erroring out sooner if this invariant is ever violated (and that error would go to Sentry, so we could actually discover it).

Reply:
Changed in 54f5e49b8af67c18aacd939b6d1e61b864b4937f to use specific types for the res.locals variant.

Reply:
That is now #12277